File: generated-code/javascript/codecraft/FileReadWrite/model/entity-action.js
const AttackAction = require.main.require('./model/attack-action');
const BuildAction = require.main.require('./model/build-action');
const MoveAction = require.main.require('./model/move-action');
const RepairAction = require.main.require('./model/repair-action');
/**
 * Entity's action
 */
class EntityAction {
    /**
     * Move action
     */
    moveAction;
    /**
     * Build action
     */
    buildAction;
    /**
     * Attack action
     */
    attackAction;
    /**
     * Repair action
     */
    repairAction;

    constructor(moveAction, buildAction, attackAction, repairAction) {
        this.moveAction = moveAction;
        this.buildAction = buildAction;
        this.attackAction = attackAction;
        this.repairAction = repairAction;
    }

    /**
     * Read EntityAction from input stream
     */
    static async readFrom(stream) {
        let moveAction;
        if (await stream.readBool()) {
            moveAction = await MoveAction.readFrom(stream);
        } else {
            moveAction = null;
        }
        let buildAction;
        if (await stream.readBool()) {
            buildAction = await BuildAction.readFrom(stream);
        } else {
            buildAction = null;
        }
        let attackAction;
        if (await stream.readBool()) {
            attackAction = await AttackAction.readFrom(stream);
        } else {
            attackAction = null;
        }
        let repairAction;
        if (await stream.readBool()) {
            repairAction = await RepairAction.readFrom(stream);
        } else {
            repairAction = null;
        }
        return new EntityAction(moveAction, buildAction, attackAction, repairAction);
    }

    /**
     * Write EntityAction to output stream
     */
    async writeTo(stream) {
        let moveAction = this.moveAction;
        if (moveAction === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await moveAction.writeTo(stream);
        }
        let buildAction = this.buildAction;
        if (buildAction === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await buildAction.writeTo(stream);
        }
        let attackAction = this.attackAction;
        if (attackAction === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await attackAction.writeTo(stream);
        }
        let repairAction = this.repairAction;
        if (repairAction === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await repairAction.writeTo(stream);
        }
    }
}
module.exports = EntityAction

File: generated-code/pascal/codecraft/FileReadWrite/Model/EntityAction.pas
unit EntityAction;

{$mode delphi}{$H+}

interface

uses
    AttackAction in 'Model/AttackAction.pas',
    BuildAction in 'Model/BuildAction.pas',
    MoveAction in 'Model/MoveAction.pas',
    Nullable,
    RepairAction in 'Model/RepairAction.pas',
    Stream,
    SysUtils;

type
    // Entity's action
    TEntityAction = class
        // Move action
        moveAction: TNullable<TMoveAction>;
        // Build action
        buildAction: TNullable<TBuildAction>;
        // Attack action
        attackAction: TNullable<TAttackAction>;
        // Repair action
        repairAction: TNullable<TRepairAction>;
        constructor Create(moveAction: TNullable<TMoveAction>; buildAction: TNullable<TBuildAction>; attackAction: TNullable<TAttackAction>; repairAction: TNullable<TRepairAction>);
        // Read EntityAction from input stream
        class function ReadFrom(stream: TStream): TEntityAction; static;
        // Write EntityAction to output stream
        procedure WriteTo(stream: TStream);
        function ToString: ansistring; override;
    end;

implementation

constructor TEntityAction.Create(moveAction: TNullable<TMoveAction>; buildAction: TNullable<TBuildAction>; attackAction: TNullable<TAttackAction>; repairAction: TNullable<TRepairAction>);
begin
    self.moveAction := moveAction;
    self.buildAction := buildAction;
    self.attackAction := attackAction;
    self.repairAction := repairAction;
end;

class function TEntityAction.ReadFrom(stream: TStream): TEntityAction;
var attackAction: TNullable<TAttackAction>;
var attackActionValue: TAttackAction;
var buildAction: TNullable<TBuildAction>;
var buildActionValue: TBuildAction;
var moveAction: TNullable<TMoveAction>;
var moveActionValue: TMoveAction;
var repairAction: TNullable<TRepairAction>;
var repairActionValue: TRepairAction;
begin
    if stream.ReadBoolean then begin
        moveActionValue := TMoveAction.ReadFrom(stream);
        moveAction := moveActionValue;
    end else
        moveAction := nil;
    if stream.ReadBoolean then begin
        buildActionValue := TBuildAction.ReadFrom(stream);
        buildAction := buildActionValue;
    end else
        buildAction := nil;
    if stream.ReadBoolean then begin
        attackActionValue := TAttackAction.ReadFrom(stream);
        attackAction := attackActionValue;
    end else
        attackAction := nil;
    if stream.ReadBoolean then begin
        repairActionValue := TRepairAction.ReadFrom(stream);
        repairAction := repairActionValue;
    end else
        repairAction := nil;
    result := TEntityAction.Create(moveAction, buildAction, attackAction, repairAction);
end;

procedure TEntityAction.WriteTo(stream: TStream);
var attackActionValue: TAttackAction;
var buildActionValue: TBuildAction;
var moveActionValue: TMoveAction;
var repairActionValue: TRepairAction;
begin
    if moveAction.HasValue then begin
        stream.WriteBoolean(true);
        moveActionValue := moveAction.Value;
        moveActionValue.WriteTo(stream);
    end else
        stream.WriteBoolean(false);
    if buildAction.HasValue then begin
        stream.WriteBoolean(true);
        buildActionValue := buildAction.Value;
        buildActionValue.WriteTo(stream);
    end else
        stream.WriteBoolean(false);
    if attackAction.HasValue then begin
        stream.WriteBoolean(true);
        attackActionValue := attackAction.Value;
        attackActionValue.WriteTo(stream);
    end else
        stream.WriteBoolean(false);
    if repairAction.HasValue then begin
        stream.WriteBoolean(true);
        repairActionValue := repairAction.Value;
        repairActionValue.WriteTo(stream);
    end else
        stream.WriteBoolean(false);
end;

function TEntityAction.ToString: ansistring;
var attackActionValue: TAttackAction;
var buildActionValue: TBuildAction;
var moveActionValue: TMoveAction;
var repairActionValue: TRepairAction;
begin
    result := 'EntityAction {';
    result += 'moveAction=';
    if moveAction.HasValue then begin
        moveActionValue := moveAction.Value;
        result += moveActionValue.ToString;
    end else
        result += 'nil';
    result += ', ';  
    result += 'buildAction=';
    if buildAction.HasValue then begin
        buildActionValue := buildAction.Value;
        result += buildActionValue.ToString;
    end else
        result += 'nil';
    result += ', ';  
    result += 'attackAction=';
    if attackAction.HasValue then begin
        attackActionValue := attackAction.Value;
        result += attackActionValue.ToString;
    end else
        result += 'nil';
    result += ', ';  
    result += 'repairAction=';
    if repairAction.HasValue then begin
        repairActionValue := repairAction.Value;
        result += repairActionValue.ToString;
    end else
        result += 'nil';
    result += '}';
end;

end.